lib/codegen.ml

```diff
@@ -0,0 +1,303 @@
+open! Base
+open Printf
+open! Ppxlib_ast
+open Parsetree
+open Asttypes
+open Longident
+open! Ast_helper
+module S = Set
+open Syntax
+open Efsm
+
+let state_action_type (g : G.t) (st : int) =
+  let merge_state_action_type aty1 aty2 =
+    match (aty1, aty2) with
+    | `Terminal, aty2 -> aty2
+    | aty1, `Terminal -> aty1
+    | `Send, `Send -> `Send
+    | `Recv, `Recv -> `Recv
+    | `Send, `Recv -> `Mixed
+    | `Recv, `Send -> `Mixed
+    | aty1, `Mixed -> aty1
+    | `Mixed, aty2 -> aty2
+  in
+  let f (_, a, _) acc =
+    let aty =
+      match a with
+      | SendA _ -> `Send
+      | RecvA _ -> `Recv
+      | Epsilon -> failwith "Impossible"
+    in
+    merge_state_action_type aty acc
+  in
+  G.fold_succ_e f g st `Terminal
+
+let mk_lid id = Location.mknoloc (parse id)
+
+let mk_constr id = Typ.constr (mk_lid id) []
+
+let loc = Location.none
+
+let unit = [%type: unit]
+
+let gen_callback_module (g : G.t) : structure_item =
```

Comment:
this function is quite long and some cases look quite similar, perhaps some helper functions can be factored out